File: packages/go/fixtures.go
// Canonical Go fixture for the V1 contract. The endpoint summaries are
// equivalent to packages/contracts/src/fixtures.ts so ingest validates both
// runtimes into the same internal event shape. Event IDs are unique per
// fixture because they represent distinct observed requests.
package apphealth

import (
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"fmt"
)

// CanonicalEndpointSummaries mirrors the TypeScript CANONICAL_ENDPOINT_SUMMARIES.
var CanonicalEndpointSummaries = []struct {
	Method     string
	Route      string
	StatusCode int
	DurationMs int
}{
	{"GET", "/health", 200, 12},
	{"GET", "/users/:id", 200, 45},
	{"GET", "/users/:id", 404, 38},
	{"POST", "/orders", 201, 120},
	{"POST", "/orders", 500, 350},
	{"GET", "/orders/:id", 200, 88},
}

const fixtureNow int64 = 1_725_000_000_000

// deterministicUUID derives an RFC 4122 v4 UUID from a seed so fixtures are stable.
func deterministicUUID(seed uint32) string {
	h := sha256.New()
	var buf [4]byte
	binary.BigEndian.PutUint32(buf[:], seed)
	h.Write(buf[:])
	sum := h.Sum(nil)
	hexStr := hex.EncodeToString(sum[:16])
	// Force version 4 and variant bits.
	hexStr = hexStr[:8] + "-" + hexStr[8:12] + "-4" + hexStr[13:16] + "-a" + hexStr[17:20] + "-" + hexStr[20:32]
	return hexStr
}

// BuildCanonicalBatch builds a v1 batch from the canonical summaries tagged
// with the given runtime. seedBase distinguishes Node and Go fixtures so
// event IDs are unique while endpoint summaries remain equivalent.
func BuildCanonicalBatch(runtime Runtime, release string, seedBase uint32) EventBatchV1 {
	events := make([]EventV1, len(CanonicalEndpointSummaries))
	for i, s := range CanonicalEndpointSummaries {
		r := release
		events[i] = EventV1{
			EventID:    deterministicUUID(seedBase + uint32(i*10)),
			Timestamp:  fixtureNow + int64(i*1000),
			Method:     s.Method,
			Route:      s.Route,
			StatusCode: s.StatusCode,
			DurationMs: s.DurationMs,
			Release:    &r,
		}
	}
	return EventBatchV1{
		BatchID:       deterministicUUID(seedBase - 1),
		SchemaVersion: SchemaVersion,
		Runtime:       runtime,
		Release:       &release,
		Events:        events,
	}
}

// NodeBatchFixture returns the canonical Node fixture built from Go so the Go
// test suite can prove endpoint equivalence with the TypeScript fixture.
func NodeBatchFixture() EventBatchV1 {
	r := "0.0.0-fixture"
	return BuildCanonicalBatch(RuntimeNode, r, 1000)
}

// GoBatchFixture returns the canonical Go fixture.
func GoBatchFixture() EventBatchV1 {
	r := "0.0.0-fixture"
	return BuildCanonicalBatch(RuntimeGo, r, 2000)
}

// EndpointKey is the normalized (method, route, status, duration) tuple used
// for fixture equivalence checks.
type EndpointKey struct {
	Method     string
	Route      string
	StatusCode int
	DurationMs int
}

// EndpointKeys returns the sorted normalized endpoint-summary keys for a batch.
func EndpointKeys(b EventBatchV1) []EndpointKey {
	keys := make([]EndpointKey, len(b.Events))
	for i, e := range b.Events {
		keys[i] = EndpointKey{e.Method, e.Route, e.StatusCode, e.DurationMs}
	}
	// Stable sort by (method, route, status, duration).
	for i := 1; i < len(keys); i++ {
		for j := i; j > 0; j-- {
			a, b := keys[j-1], keys[j]
			if a.Method < b.Method ||
				(a.Method == b.Method && a.Route < b.Route) ||
				(a.Method == b.Method && a.Route == b.Route && a.StatusCode < b.StatusCode) ||
				(a.Method == b.Method && a.Route == b.Route && a.StatusCode == b.StatusCode && a.DurationMs < b.DurationMs) {
				keys[j-1], keys[j] = keys[j], keys[j-1]
			} else {
				break
			}
		}
	}
	return keys
}

// AreEndpointEquivalent returns true if two batches carry equivalent endpoint
// summaries (method, route, status, duration), ignoring event IDs and timestamps.
func AreEndpointEquivalent(a, b EventBatchV1) bool {
	ka, kb := EndpointKeys(a), EndpointKeys(b)
	if len(ka) != len(kb) {
		return false
	}
	for i := range ka {
		if ka[i] != kb[i] {
			return false
		}
	}
	return true
}

// String returns a debug string for an EndpointKey.
func (k EndpointKey) String() string {
	return fmt.Sprintf("%s %s %d %dms", k.Method, k.Route, k.StatusCode, k.DurationMs)
}

File: packages/contracts/src/fixtures.ts
// Canonical Node, Cloudflare Worker, and Go contract fixtures for V1.
// All fixtures carry equivalent endpoint summaries (same method, route,
// status_code, duration_ms, and release) so ingest validates them into the
// same internal event shape. Event IDs are unique per fixture because they
// represent distinct observed requests; equivalence is asserted on the
// endpoint-summary fields, not on event_id.

import type { EventBatchV1 } from './event.js';

const NOW = 1_725_000_000_000; // deterministic fixture timestamp

function uuid(seed: number): string {
  // Deterministic RFC 4122 v4 UUID derived from a seed so fixtures are stable.
  const hex = (n: number, len: number): string =>
    Math.abs(Math.floor(n * 2654435761))
      .toString(16)
      .padStart(len, '0')
      .slice(-len);
  return (
    `${hex(seed, 8)}-${hex(seed + 1, 4).slice(0, 4)}-4${hex(seed + 2, 3)}` +
    `-a${hex(seed + 3, 3)}-${hex(seed + 4, 8)
      .padStart(12, '0')
      .slice(-12)}`
  );
}

export interface EndpointSummary {
  method: string;
  route: string;
  status_code: number;
  duration_ms: number;
}

/** Canonical endpoint summaries shared by both Node and Go fixtures. */
export const CANONICAL_ENDPOINT_SUMMARIES: readonly EndpointSummary[] = [
  { method: 'GET', route: '/health', status_code: 200, duration_ms: 12 },
  { method: 'GET', route: '/users/:id', status_code: 200, duration_ms: 45 },
  { method: 'GET', route: '/users/:id', status_code: 404, duration_ms: 38 },
  { method: 'POST', route: '/orders', status_code: 201, duration_ms: 120 },
  { method: 'POST', route: '/orders', status_code: 500, duration_ms: 350 },
  { method: 'GET', route: '/orders/:id', status_code: 200, duration_ms: 88 },
] as const;

/** Build a v1 batch from canonical summaries tagged with a runtime. */
export function buildCanonicalBatch(
  runtime: 'node' | 'worker' | 'go',
  release = '0.0.0-fixture',
  seedBase = runtime === 'node' ? 1000 : runtime === 'worker' ? 1500 : 2000,
): EventBatchV1 {
  return {
    batch_id: uuid(seedBase - 1),
    schema_version: 'v1',
    runtime,
    release,
    events: CANONICAL_ENDPOINT_SUMMARIES.map((summary, index) => ({
      event_id: uuid(seedBase + index * 10),
      timestamp: NOW + index * 1000,
      method: summary.method,
      route: summary.route,
      status_code: summary.status_code,
      duration_ms: summary.duration_ms,
      release,
    })),
  };
}

/** Canonical Node fixture. */
export const nodeBatchFixture = (): EventBatchV1 => buildCanonicalBatch('node');

/** Canonical Cloudflare Worker fixture. */
export const workerBatchFixture = (): EventBatchV1 => buildCanonicalBatch('worker');

/** Canonical Go fixture. */
export const goBatchFixture = (): EventBatchV1 => buildCanonicalBatch('go');

/**
 * Two batches are endpoint-equivalent if, after sorting events by
 * (method, route, status_code, duration_ms), their normalized summaries match.
 * Event IDs and timestamps are intentionally ignored.
 */
export function areEndpointEquivalent(a: EventBatchV1, b: EventBatchV1): boolean {
  const norm = (batch: EventBatchV1) =>
    batch.events
      .map((e) => `${e.method}|${e.route}|${e.status_code}|${e.duration_ms}`)
      .sort()
      .join('\n');
  return norm(a) === norm(b);
}
